File: PallindromeList.java
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
        next = null;
    }
}

public class PallindromeList {

    // TC: O(n) n - number of nodes in a linked list
    // SC: O(1)
    public boolean isPalindrome(ListNode head) {
        if(head == null) return true;

        // 1. Find out the middle node
        ListNode middleNode = getMiddleNode(head);

        // 2. reverse whatever is there after the middle node
        ListNode secondHead = reverseList(middleNode.next);
        middleNode.next = null;

        ListNode firstHead = head;

        // 3. compare the nodes one by one
        while(secondHead != null) {
            // If two pointers match
            if(firstHead.val == secondHead.val) {
                firstHead = firstHead.next;
                secondHead = secondHead.next;
            }else {
                return false;
            }

        }
        return true;
    }

    private ListNode getMiddleNode(ListNode head) {

        ListNode slow = head;
        ListNode fast = head;

        while(fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    private ListNode reverseList(ListNode head) {
        if(head == null || head.next == null) return head;

        ListNode prev = null;
        ListNode current = head;
        ListNode next = head.next;


        while(next != null) {
            current.next = prev;
            prev = current;
            current = next;
            next = next.next;
        }

        current.next = prev;

        return current;
    }
}

File: pallindromeList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
'''
Time Complexity -->
O(n)

Space complexity -->

O(1)
'''
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = head
        if slow.next==None:
            return True
        if slow.next.next==None:
            if slow.val == slow.next.val:
                return True
            else:
                return False
        
        middle = None
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        middle = slow
        #print(middle.val)
        middle.next = self.reverseList(middle)
        #print(middle.next.val)
        
        curr2 = middle.next
        curr1 = head
        while curr2:
            if curr1.val !=curr2.val:
                return False
            curr1 = curr1.next
            curr2 = curr2.next
        return True
    
    def reverseList(self, node):
        prev = None
        curr = node.next
        nextNode = node.next.next
        #print(nextNode.val)
        while nextNode:
            curr.next = prev
            prev = curr
            curr = nextNode
            nextNode = nextNode.next
        curr.next = prev
        prev = curr
        #print(prev.val)
        return prev
            
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
            
            
        